### ISSUE TITLE
vllm0.17.0 版本测试，没有收到性能收益

### ISSUE THREAD

**Issue by chenjiaoAngel** [qhfan/FlashPrefill#7]
测试脚本：
`
# from patches import patch_loader_vllm_0_10_0
# from patches import patch_loader_vllm_0_12_0
from patches import patch_loader_vllm_0_17_0
from vllm import LLM, SamplingParams
# model_name="/mnt/bn/bandai-lf/shared/models/Qwen3.5-35B-A3B"
model_name="/mnt/bn/bandai-lf/shared/models/Qwen3-30B-A3B-Base"
import  time
def test_generate(model, sampling_params, prompt, warmup_times=1, repeats=10):
    # Warmup阶段
    print(f"开始Warmup, 共执行{warmup_times}次...")
    for i in range(warmup_times):
        start_time = time.time()
        outputs = model.generate(prompts=[prompt], sampling_params=sampling_params)
        end_time = time.time()
        print(f"Warmup {i+1}/{warmup_times}，耗时: {end_time - start_time:.4f}秒")
    
    # 正式测试阶段
    print(f"\n开始正式测试, 共执行{repeats}次...")
    total_time = 0
    for i in range(repeats):
        start_time = time.time()
        outputs = model.generate(prompts=[prompt], sampling_params=sampling_params)
        end_time = time.time()
        current_time = end_time - start_time
        total_time += current_time
        print(f"测试 {i+1}/{repeats}，耗时: {current_time:.4f}秒")
    print(f"\n测试完成!")
    print(f"总耗时: {total_time:.4f}秒")
    avg_time = total_time / repeats
    print(f"平均耗时: {avg_time:.4f}秒")
    
    # 获取最后一次的输出结果
    output = outputs[0].outputs[0].text
    return avg_time, output

if __name__ == "__main__":
    model = LLM(
        model=model_name,
        tokenizer=model_name,
        trust_remote_code=True,
        dtype="bfloat16",
        tensor_parallel_size=4,
        max_model_len=32 * 1024, 
        gpu_memory_utilization=0.9,
        enable_chunked_prefill=False,
        enable_prefix_caching=False
    )
    sampling_params = SamplingParams(temperature=0, max_tokens=64)
    PROMPT = "请详细介绍大模型的分布式推理技术，从张量并行、流水线并行、分页注意力三个核心方向展开，要求内容详实、逻辑清晰，总字数不少于20000字。"
    outputs=model.generate(prompts=[PROMPT], sampling_params=sampling_params)
    output = outputs[0].outputs[0].text
    print(output)
    avg_time, output = test_generate(model, sampling_params, PROMPT)
    print("\n生成结果:")
    print(output)
    print(f"平均耗时: {avg_time:.4f}秒")
`

vllm.0.17.0 patch ：

`
import os
import torch
import types
# from vllm_models.ops import flash_sparse_infer_out
from ops import flash_prefill_varlen_func, flash_prefill
# from vllm_models.ops.flashparseinfernotune import flash_sparse_infer_out
# import vllm.model_executor.model_loader as vllm_loader
from vllm.v1.attention.backends.flash_attn import FlashAttentionMetadata, cascade_attention, FlashAttentionBackend
from vllm.v1.attention.backend import AttentionType
from vllm.vllm_flash_attn import  flash_attn_varlen_func
from vllm._custom_ops import reshape_and_cache_flash

ATTENTION_CONFIG_EXAMPLE = {
    "default": {},
    "flash": {},
    "flashprefill_llama": {
        "block_size": 128,
        "attention_sink": 2,
        "window_size": 4,
        "alpha": 0.18,
        "last_n_full_block": 2,
        # "min_budget": 4
    },
    "flashprefill_qwen2": {
        "block_size": 128,
        "attention_sink": 2,
        "window_size": 4,
        "alpha": 0.08,
        "last_n_full_block": 2,
        # "min_budget": 4
    },
    "flashprefill_qwen3moe": {
        "block_size": 128,
        "attention_sink": 2,
        "window_size": 4,
        "alpha": 0.12,
        "last_n_full_block": 2,
        # "min_budget": 4
    }
}

# use qwen2_5 as example
cfg = ATTENTION_CONFIG_EXAMPLE["flashprefill_qwen3moe"]

def apply_patch_to_worker(worker=None):
    print("====apply_patch_to_worker====")
    import vllm.v1.attention.backends.flash_attn as fa_backend

    rank = str(getattr(worker, "rank", os.environ.get("RANK", "0")))
    cache_dir = f"/tmp/triton_cache_rank_{rank}"
    os.makedirs(cache_dir, exist_ok=True)
    os.environ["TRITON_CACHE_DIR"] = cache_dir

    print(cache_dir)

    original_forward = fa_backend.FlashAttentionImpl.forward

    def flash_prefill_forward(
        self,
        layer: torch.nn.Module,
        query: torch.Tensor,
        key: torch.Tensor,
        value: torch.Tensor,
        kv_cache: torch.Tensor,
        attn_metadata: FlashAttentionMetadata,
        output: torch.Tensor | None = None,
        output_scale: torch.Tensor | None = None,
        output_block_scale: torch.Tensor | None = None,
    ) -> torch.Tensor:
        """Forward pass with FlashAttention.

        Args:
            query: shape = [num_tokens, num_heads, head_size]
            key: shape = [num_tokens, num_kv_heads, head_size]
            value: shape = [num_tokens, num_kv_heads, head_size]
            kv_cache: shape =
                [2, num_blocks, block_size, num_kv_heads, head_size]
            attn_metadata: Metadata for attention.
        Returns:
            shape = [num_tokens, num_heads * head_size]
        NOTE: FP8 quantization, flash-attn expect the size of
              {q,k,v}_descale to be (num_sequences, num_kv_heads).
              We use torch's .expand() to avoid duplicating values
        """
        assert output is not None, "Output tensor must be provided."

        if output_scale is not None or output_block_scale is not None:
            raise NotImplementedError(
                "fused output quantization is not yet supported for FlashAttentionImpl"
            )

        if attn_metadata is None:
            # Profiling run.
            return output.fill_(0)
        
        attn_type = self.attn_type

        # IMPORTANT!
        # NOTE(woosuk): With piece-wise CUDA graphs, this method is executed in
        # eager-mode PyTorch. Thus, we need to be careful about any CPU overhead
        # in this method. For example, `view` and `slice` (or `[:n]`) operations
        # are surprisingly slow even in the case they do not invoke any GPU ops.
        # Minimize the PyTorch ops in this method as much as possible.
        # Whenever making a change in this method, please benchmark the
        # performance to make sure it does not introduce any overhead.

        num_actual_tokens = attn_metadata.num_actual_tokens

        # Handle encoder attention differently - no KV cache needed
        if attn_type in (AttentionType.ENCODER_ONLY, AttentionType.ENCODER):
            # For encoder attention,
            # we use direct Q, K, V tensors without caching
            return self._forward_encoder_attention(
                query[:num_actual_tokens],
                key[:num_actual_tokens],
                value[:num_actual_tokens],
                output[:num_actual_tokens],
                attn_metadata,
                layer,
            )

        # For decoder and cross-attention, use KV cache as before
        key_cache, value_cache = kv_cache.unbind(0)

        # key and value may be None in the case of cross attention. They are
        # calculated once based on the output from the encoder and then cached
        # in KV cache.
        if (
            self.kv_sharing_target_layer_name is None
            and key is not None
            and value is not None
        ):
            # Reshape the input keys and values and store them in the cache.
            # Skip this if sharing KV cache with an earlier attention layer.
            # NOTE(woosuk): Here, key and value are padded while slot_mapping is
            # not padded. However, we don't need to do key[:num_actual_tokens]
            # and value[:num_actual_tokens] because the reshape_and_cache_flash
            # op uses the slot_mapping's shape to determine the number of
            # actual tokens.
            reshape_and_cache_flash(
                key,
                value,
                key_cache,
                value_cache,
                attn_metadata.slot_mapping,
                self.kv_cache_dtype,
                layer._k_scale,
                layer._v_scale,
            )

        if self.kv_cache_dtype.startswith("fp8"):
            # queries are quantized in the attention layer
            dtype = FlashAttentionBackend.get_fp8_dtype_for_flashattn(
                self.kv_cache_dtype
            )
            key_cache = key_cache.view(dtype)
            value_cache = value_cache.view(dtype)

        if not attn_metadata.use_cascade:
            cu_seqlens_q = attn_metadata.query_start_loc
            seqused_k = attn_metadata.seq_lens
            max_seqlen_q = attn_metadata.max_query_len
            max_seqlen_k = attn_metadata.max_seq_len
            block_table = attn_metadata.block_table
            scheduler_metadata = attn_metadata.scheduler_metadata

            descale_shape = (cu_seqlens_q.shape[0] - 1, self.num_kv_heads)

            is_prefill = attn_metadata.max_query_len > 1

            if self.dcp_world_size > 1:
                self._forward_with_dcp(
                    query[:num_actual_tokens],
                    key[:num_actual_tokens],
                    value[:num_actual_tokens],
                    key_cache,
                    value_cache,
                    output[:num_actual_tokens],
                    attn_metadata,
                    q_descale=layer._q_scale.expand(descale_shape),
                    k_descale=layer._k_scale.expand(descale_shape),
                    v_descale=layer._v_scale.expand(descale_shape),
                )
                return output
            else:
                if is_prefill:
                    flash_prefill_varlen_func(
                        query[:num_actual_tokens],
                        key[:num_actual_tokens],
                        value[:num_actual_tokens],
                        output[:num_actual_tokens],
                        cu_seqlens_q,
                        max_seqlen_q,
                        cfg["alpha"],
                        cfg["block_size"],
                        cfg["attention_sink"],
                        cfg["window_size"],
                        cfg["last_n_full_block"]
                    )
                    return output
                else:
                    flash_attn_varlen_func(
                        q=query[:num_actual_tokens],
                        k=key_cache,
                        v=value_cache,
                        out=output[:num_actual_tokens],
                        cu_seqlens_q=cu_seqlens_q,
                        max_seqlen_q=max_seqlen_q,
                        seqused_k=seqused_k,
                        max_seqlen_k=max_seqlen_k,
                        softmax_scale=self.scale,
                        causal=attn_metadata.causal,
                        alibi_slopes=self.alibi_slopes,
                        window_size=self.sliding_window,
                        block_table=block_table,
                        softcap=self.logits_soft_cap,
                        scheduler_metadata=scheduler_metadata,
                        fa_version=self.vllm_flash_attn_version,
                        q_descale=layer._q_scale.expand(descale_shape),
                        k_descale=layer._k_scale.expand(descale_shape),
                        v_descale=layer._v_scale.expand(descale_shape),
                        num_splits=attn_metadata.max_num_splits,
                        s_aux=self.sinks,
                    )
                    return output

        cascade_attention(
            output[:num_actual_tokens],
            query[:num_actual_tokens],
            key_cache,
            value_cache,
            cu_query_lens=attn_metadata.query_start_loc,
            max_query_len=attn_metadata.max_query_len,
            cu_prefix_query_lens=attn_metadata.cu_prefix_query_lens,
            prefix_kv_lens=attn_metadata.prefix_kv_lens,
            suffix_kv_lens=attn_metadata.suffix_kv_lens,
            max_kv_len=attn_metadata.max_seq_len,
            softmax_scale=self.scale,
            alibi_slopes=self.alibi_slopes,
            sliding_window=self.sliding_window,
            logits_soft_cap=self.logits_soft_cap,
            block_table=attn_metadata.block_table,
            common_prefix_len=attn_metadata.common_prefix_len,
            max_num_splits=attn_metadata.max_num_splits,
            fa_version=self.vllm_flash_attn_version,
            prefix_scheduler_metadata=attn_metadata.prefix_scheduler_metadata,
            suffix_scheduler_metadata=attn_metadata.scheduler_metadata,
            q_descale=layer._q_scale,
            k_descale=layer._k_scale,
            v_descale=layer._v_scale,
            s_aux=self.sinks,
        )
        return output
    
    fa_backend.FlashAttentionImpl.forward = flash_prefill_forward
    print(f"Successfully patched Rank {rank}")

apply_patch_to_worker()
`

测试的时候通过注释`from patches import patch_loader_vllm_0_17_0` 决定是否开启flash-prefill，测试发现不开启性能要好些
不添加patch，其耗时均值：

<img width="676" height="262" alt="Image" src="https://github.com/user-attachments/assets/07bed051-e76a-4b8f-af1f-fb710369bdba" />

添加patch，其耗时均值：

<img width="698" height="228" alt="Image" src="https://github.com/user-attachments/assets/40905ab3-7899-4a53-b589-43ad5b45290c" />

测试qwen3.5-35B +128K/64K 也是没有收益

**Comment by qhfan** [comment]
该方法主要面向long context的prefill，加速的是long context阶段的prefill。输入prompt > 16K的时候才会有明显受益，并不会加速生成token。

**Comment by Rugu7** [comment]
vllm 中 is_prefill 条件是否不严格，当前阶段判断  is_prefill = attn_metadata.max_query_len > 1 ，如混合批次时，导致decode阶段会使用该算子。 @qhfan

**Comment by qhfan** [comment]
该算子没有支持decode

**Comment by chenjiaoAngel** [comment]
> 该算子没有支持decode

未来会考虑支持吗？我发现长序列128K 的rollout，大头耗时在decode

**Comment by Rugu7** [comment]
> 该算子没有支持decode

vllm版本，压测性能时，加打印 发现默认所有阶段都走的 is_prefill 阶段。

**Comment by chenjiaoAngel** [comment]
单独测试profile 和 decode 耗时，发现开启这个功能，性能还差了一倍：

<img width="1314" height="410" alt="Image" src="https://github.com/user-attachments/assets/f37753d7-04ca-45ea-afe3-e7b1228f3164" /> （没有开启的， prefill 130+）

<img width="1884" height="634" alt="Image" src="https://github.com/user-attachments/assets/9508ddd0-b12b-425f-af4e-5b21ff896b55" /> （开启的, prefill 200+）

**Comment by qhfan** [comment]
代码内包含autotune自动搜索tile_size等最优config，这一功能耗时较多，可以搜到最优配置后手动注释掉。

**Comment by qhfan** [comment]
或者直接写死这些config

**Comment by qhfan** [comment]
41个token应该太短了，我不确定有没有加速，我们最短测试是在4K的prompt上进行的

**Comment by chenjiaoAngel** [comment]
> 41个token应该太短了，我不确定有没有加速，我们最短测试是在4K的prompt上进行的

这个是20K的；128K 后面也跑了，时间变短了；但是也没有不开启的性能好

**Comment by qhfan** [comment]
可能vLLM版本变了的原因？我们最高测试到0.12.0，就算是GEMM很重的模型，在32K以后加速应该比较明显了

<img width="674" height="470" alt="Image" src="https://github.com/user-attachments/assets/313c8aef-d92b-4ab4-bd3d-3ff0d2b89bca" />

**Comment by chenjiaoAngel** [comment]
> 到0.12.0，就算是GEMM很重的模型，在32K以后加速

嗯，了解

**Comment by qhfan** [comment]
哦，我觉得应该是没注释掉autotune导致的，这个在128K时候搜索会花掉大量的时间
可以先运行一遍，在相同长度下搜索出一个最优配置(tile_size, num_warp)，然后再进行测试，我们论文里的结果测试多次会忽略掉第一次的数据
实际使用时候可以直接注释掉autotune

<img width="660" height="237" alt="Image" src="https://github.com/user-attachments/assets/ee8b2cdc-42e9-466c-a081-51d3f6f3a367" />

**Comment by chenjiaoAngel** [comment]
这个不应是是warmup 的时候，选择最优的配置；然后repeat 运行时，直接使用吗？ 我的测试有warmup 预热，然后在测试repeat 次，取均值
